source/common/config/grpc_mux_impl.cc

```diff
@@ -163,7 +163,12 @@ void GrpcMuxImpl::onReceiveMessage(std::unique_ptr<envoy::api::v2::DiscoveryResp
     return;
   }
   if (api_state_[type_url].watches_.empty()) {
-    ENVOY_LOG(warn, "Ignoring unwatched type URL {}", type_url);
+    if (message->resources().empty()) {
+      api_state_[type_url].request_.set_version_info(message->version_info());
+      api_state_[type_url].request_.set_response_nonce(message->nonce());
```

Comment:
Thanks for the fix PR @yuval-k.

Shouldn't we also do a `sendDiscoveryRequest()` (as done at EOF) to indicate back to the management server that this empty list has now been accepted? Similarly, with the two branches (above and below) with `ENVOY_LOG(warn,..`, I think we should do a `sendDiscoveryRequest()` with the new nonce but without updating version, to NACK.

Reply:
as far as I understand it, if the list of watches is empty you probably dont want to send a discovery request.
e.g. when a cluster removed, and there are no more EDS resources, no need to send a discover request. thoguhts?

In the mean time i'll update the other code paths.

Reply:
Not sure whats the best way to handle unknown type URL. sounds like something that should never happen... maybe RELEASE_ASSERT?

Reply:
Yep, you are right, there is no need to `sendDiscoveryRequest()` when the watch list is empty. On the two WARNING paths, we can't `RELEASE_ASSERT`, Envoy shouldn't crash because of a bad management server sequence. There are various behaviors that would be reasonable:
1. NACK, since this `DiscoveryResponse` could be be applied
2. Drop stream since there is a protocol violation, with a warning log.
3. Just print a warning log (existing behavior), there is a protocol violation, all bets are off.
I think (1) is maybe overkill, so either (2) or (3). Maybe just add a `TODO` at these sites to consider dropping the stream.

Reply:
Also, please add some brief comments capturing the above chain of thought, thanks!

Reply:
so to summarize:
no watches AND no resources ⇒ accept version; don't send discovery request
no watches AND some resources ⇒ reject version (update nonce but not version); send discovery request as a NACK
unknown type ⇒ Probably drop stream; add TODO for now. should 'warn' log be 'error'?